Quick solution for ssr errors

miguelchico opened the pull request:
Hi,

This is my quick solution for SSR issues. This must close issue #64 
Let me know what do you thing about it.

omridevk commented:
@miguelchico 
Great job man, much appreciated!!! 
two questions:
1. What happens after the server side rendering, do we still have access later to determine which OS it is? as I assume SSR can be used to deliver apps for both Mac and Windows alike (unless I am mistaken here), so I am wonder how will this solution works on both on OS, Mac and Windows? 
Do we need to delay the check until we actually rendered in the browser? sorry if these questions sounds stupid but my experience with Angular Universal is non-existence. 
2. What changes were made polyfills.ts? Not sure I can see what changed exactly, was it added by mistake by prettier? 

And again, thank you so much for taking your time and effort into resolving the issue.
I will merge and release a version but I would like to verify consistent behavior across platforms.

omridevk commented on projects/ng-keyboard-shortcuts/src/lib/keys.ts:
```diff
@@ -1,4 +1,7 @@
-const isMac = (navigator.userAgent.indexOf('Mac OS') !== -1);
+let isMac = false;
+if (typeof navigator !== 'undefined') {
+    (navigator.userAgent.indexOf('Mac OS') !== -1);
```
isMac is never assigned again, or unless I am missing something. 
```suggestion
const  isMac = typeof navigator !== 'undefined' ? (navigator.userAgent.includes("Mac OS") : false;
```
though it still raises the question why we decide it is not Mac because navigator is not available at the current stage.

omridevk replied:
Also I think we use new Typescript which supports:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html
which might makes the code a bit cleaner. just a suggestion, feel free to ignore.

miguelchico replied:
I applied your first suggestion

miguelchico commented:
@omridevk 
About your questions:
1. Using SSR the javascript code will be executed in both places, server and browser. The problem with objects like navigator or window or similar is they are not available in a server context so with the 'tricks' I added on the first run of the code in the server nothing will break and then once the code is again evaluated in the browser the previous functionality will be present again.
2. I just added an if block
`if (typeof Element !== 'undefined') {`
around the usages of Element.prototype. I think compare messed it a little bit.

omridevk commented on projects/ng-keyboard-shortcuts/src/lib/keys.ts:
```diff
@@ -1,6 +1,6 @@
 let isMac = false;
 if (typeof navigator !== 'undefined') {
```
can be removed.

omridevk replied:
extra space between, weird that prettier didn't fix it. maybe because we used the apply suggestion.

omridevk commented:
@miguelchico 
Thank you for your answers
1. I understand that obviously navigator object is not available in SSR context, but I was just wondering how and when the code is evaluated again after being served to the browser.
and may I ask if you tested it both in iOS and windows? just to make sure the behavior is consistent.
sorry  for the stupid question, I never got a chance to actually play with SSR in Angular context. 
I am not afraid of regressions here since this code will not affect normal angular built.
2. just wondering, Element object is not available in SSR context as well?

omridevk commented:
@miguelchico 
How about this for polyfills.ts:
let me know what you think:
```typescript
const initElement = (function() {
    if (typeof Element === "undefined") {
        return;
    }
    if (!Element.prototype.matches) {
        Element.prototype.matches =
            (Element.prototype as any).msMatchesSelector || Element.prototype.webkitMatchesSelector;
    }

    if (!Element.prototype.closest) {
        Element.prototype.closest = function(s) {
            let el = this;

            do {
                if (el.matches(s)) return el;
                el = el.parentElement || el.parentNode;
            } while (el !== null && el.nodeType === 1);
            return null;
        };
    }
})();

if (!(Array.prototype as any).flat) {
    (Array.prototype as any).flat = function(depth) {
        var flattend = [];
        (function flat(array, depth) {
            for (let el of array) {
                if (Array.isArray(el) && depth > 0) {
                    flat(el, depth - 1);
                } else {
                    flattend.push(el);
                }
            }
        })(this, Math.floor(depth) || 1);
        return flattend;
    };
}

if (!(Array.prototype as any).flatMap) {
    (Array.prototype as any).flatMap = function() {
        return Array.prototype.map.apply(this, arguments).flat(1);
    };
}

```

seems to solve the issue, and I am more of a fan of defensing programming, and early escapes,
let me know what you think please.

omridevk commented:
and similar for keys.ts:
```typescript
const isMac = typeof navigator !== "undefined" ? navigator.userAgent.includes("Mac OS") : false;
```

omridevk commented:
anyway, you did a great job man, I'll merge it and release a new version.
Thank you very much for contribution.

miguelchico commented:
> @miguelchico
> Thank you for your answers
> 
> 1. I understand that obviously navigator object is not available in SSR context, but I was just wondering how and when the code is evaluated again after being served to the browser.
>    and may I ask if you tested it both in iOS and windows? just to make sure the behavior is consistent.
>    sorry  for the stupid question, I never got a chance to actually play with SSR in Angular context.
>    I am not afraid of regressions here since this code will not affect normal angular built.
> 2. just wondering, Element object is not available in SSR context as well?

@omridevk 
1. The code is first evaluated in a server side (Linux, macos, windows, it doesn't matter) to be able to provider a first page with all the HTML and CSS already parsed to help the browser to do a fast first rendering of your application. Besides the HTML, CSS, the regular JS files are linked to the page and they are parsed in the browser. So it doesn't matter if you serve your app in macos, linux or windows because the library will be parsed again in any browser.

2. Because Element is an object to modify HTML elements, it doesn't make sense in a pure server side and that's why it's not available (I think)

miguelchico commented:
> anyway, you did a great job man, I'll merge it and release a new version.
> Thank you very much for contribution.

I just tested the base application with the new version and everything is working like a charm!

Thank you!!!
